### ISSUE TITLE
[Feature Request] OAuth authentication

### ISSUE THREAD

**Issue by Steeven9** [sbondCo/Watcharr#461]
The docs mention that you can log in with Jellyfin, but why not take this one step further and allow to register your own OAuth 2.0 provider for SSO?
For example, one might have a local Keycloak instance set up, or want to use their Google account.

**Comment by IRHM** [comment]
Hey @Steeven9, thanks for opening this issue!

We can definitely look into this at some point, there's quite a bit of other tasks ahead of this, so I'm not sure when I'll be able to look at it. If you or someone else would like to, feel free!

I have never hosted/used my own instance of a oauth/sso provider, hopefully they are all the same so we can just add generic setting that you could use for any service.

**Comment by Steeven9** [comment]
Thanks for the quick reply! I unfortunately have never implemented an SSO login, only _configured_ a few so far :)

Thankfully OAuth/OIDC is a pretty vastly used and well-documented standard so it shouldn't be too hard to find some libraries/examples to implement it, for example https://authjs.dev

**Comment by mgrimace** [comment]
I'm jumping in to speak in support of this, specifically for [Authentik](https://goauthentik.io) support. Ideally, to allow reverse proxy authentication by header. 

I have setup NPM proxy, with Authentik for authentication sitting before Watcharr and other services. So every time I open Watcharr's webui, NPM redirects it to Authentik for authentication. Authentik can use Plex as an SSO, and would ideally pass on the user and other related proxy authentication headers to Watcharr. If there is already a user with the same username, Watcharr would then automatically login in the user without requiring it's own login screen.

The benefit of this is that I can use secure athentication service, and I wouldn't have to login twice - if that makes sense.

AFAIK Authentik uses the header `X-authentik-username`, which is how I accomplish this in Calibre-web.

**Comment by lufixSch** [comment]
I would also really appreciate OAuth/OIDC support, but I understand this might be more complicated (I also haven't implemented it before)
An alternative that could be forward auth as mentioned by @mgrimace. This should be as far as I understand relatively straight forward to implement.

**Comment by royaltongue** [comment]
Is this still being considered? I'd also appreciate being able to have my family log in with OIDC through Authentik

**Comment by lufixSch** [comment]
@royaltongue Forward Auth/Proxy Authentication is implemented and works well for me using Authentik with Traefik as a reverse proxy.

OIDC is not implemented.

**Comment by mgrimace** [comment]
> [@royaltongue](https://github.com/royaltongue) Forward Auth/Proxy Authentication is implemented and works well for me using Authentik with Traefik as a reverse proxy.
> 
> OIDC is not implemented.

~~Would you be willing to share some quick novice-friendly instruction for setting this up with Authentik? e.g, in the app (header name), and anything special in authentik or just a usual forwardauth proxy?~~

Scratch that, I added X-authentik-user, now I have 3 users with the same username: mgrimace (watcharr login), mgrimace (plex login), and mgrimace (proxy login). Is there a way to merge or make them the same user in the database by username?

**Comment by mgrimace** [comment]
Yes, confirming that both X-authentik-username and X-authentik-email both create new accounts even if there is an existing username or email login with the same credentials. Using Authentik and NPM as reverse proxy.

**Comment by lufixSch** [comment]
@mgrimace That each login method creates a separate user is intended. You can not merge users, but you can convert users from one login type to another (i.e.: from _Watcharr_ to _Proxy_) in the "Users" panel if you are logged in as an admin.

**Comment by ornelasf1** [comment]
It's worth mentioning that I'm getting the benefits of SSO by using Jellyfin's LDAP plugin https://github.com/jellyfin/jellyfin-plugin-ldapauth. Since Jellyfin authenticates with my LDAP database, I can simply select the Jellyfin login in Watcharr.

Now, it would be nice if we can remove the Watcharr login option. I'll likely create a feature request for that
